tests/integrationv2/test_external_psk.py

```diff
@@ -0,0 +1,213 @@
+import copy
+import os
+import pytest
+import time
+
+from configuration import available_ports
+from common import ProviderOptions, Protocols, data_bytes, Ciphers, Certificates
+from fixtures import managed_process
+from providers import S2N, OpenSSL
+from utils import invalid_test_parameters, get_parameter_name
+
+# Test Vectors from https://tools.ietf.org/html/rfc8448#section-4 
+shared_psk_identity = '2c035d829359ee5ff7af4ec900000000262a6494dc486d2c8a34cb33fa90bf1b00'\
+                      '70ad3c498883c9367c09a2be785abc55cd226097a3a982117283f82a03a143efd3'\
+                      'ff5dd36d64e861be7fd61d2827db279cce145077d454a3664d4e6da4d29ee03725'\
+                      'a6a4dafcd0fc67d2aea70529513e3da2677fa5906c5b3f7d8f92f228bda40dda72'\
+                      '1470f9fbf297b5aea617646fac5c03272e970727c621a79141ef5f7de6505e5bfb'\
+                      'c388e93343694093934ae4d357'
+
+shared_psk_secret = '4ecd0eb6ec3b4d87f5d6028f922ca4c5851a277fd41311c9e62d2c9492e1c4f3'
+
+s2n_known_value_psk = '2c035d829359ee5ff7af4ec900000000262a6494dc486d2c8a34cb33fa90bf1b00'\
+                      '70ad3c498883c9367c09a2be785abc55cd226097a3a982117283f82a03a143efd3'\
+                      'ff5dd36d64e861be7fd61d2827db279cce145077d454a3664d4e6da4d29ee03725'\
+                      'a6a4dafcd0fc67d2aea70529513e3da2677fa5906c5b3f7d8f92f228bda40dda72'\
+                      '1470f9fbf297b5aea617646fac5c03272e970727c621a79141ef5f7de6505e5bfb'\
+                      'c388e93343694093934ae4d357,'\
+                      '4ecd0eb6ec3b4d87f5d6028f922ca4c5851a277fd41311c9e62d2c9492e1c4f3,'\
+                      'S2N_PSK_HMAC_SHA256'
+
+# Arbitrary Test vectors
+s2n_client_only_psk = 's2n_client_psk_identity,'\
+                      'a6dadae4567876,'\
+                      'S2N_PSK_HMAC_SHA384'\
+
+s2n_server_only_psk = 's2n_server_psk_identity,'\
+                      'a64dafcd0fc67d2a,'\
+                      'S2N_PSK_HMAC_SHA256'\
+
+s2n_shared_psk = 's2n_psk_identity,'\
+                 'f9bf29b5aea6aea7,'\
+                 'S2N_PSK_HMAC_SHA256'\
+
+s2n_invalid_hmac_psk = 'psk_identity,'\
+                       'f9bf29b5aea6aea7,'\
+                       'S2N_PSK_HMAC_SHA512'\
+
+S2N_CLIENT_PSK_PARAMETERS = [ 
+    [ '--psk', s2n_client_only_psk ],
+    [ '--psk', s2n_known_value_psk ], 
+    [ '--psk', s2n_invalid_hmac_psk ],
+    [ '--psk', s2n_client_only_psk, '--psk', s2n_shared_psk ], 
+]
+
+S2N_SERVER_PSK_PARAMETERS = [
+    [ '--psk', s2n_server_only_psk ],
+    [ '--psk', s2n_known_value_psk ],
+    [ '--psk', s2n_invalid_hmac_psk ],
+    [ '--psk', s2n_server_only_psk, '--psk', s2n_shared_psk ], 
+]
+
+OPENSSL_PSK_PARAMETER = [ '-psk_identity', shared_psk_identity, '--psk', shared_psk_secret ]
+
+def s2n_assert_chosen_psk(idx, results, random_bytes, is_s2nd=False, is_peer_openssl_server=False):
+    if idx == 0:
```

Comment:
Determining the correct results this way is extremely confusing. If checking the results have nothing in common, why do it in a common method?

But shouldn't you only have two types of results? Either we negotiate PSK with a specific identity, or we do a full handshake.

Reply:
This is the case for `s2nd` as it would fallback to full handshake if psk is not negotiated, in case of `openssl s_server` however, it throws an error and does not fallback to a full handshake.